source/key.ts

```diff
@@ -0,0 +1,366 @@
+            return importPrivateKeyFromPKCS8(key.priv);
+        case D4LKeyTypes.USER_PUBLIC_KEY:
+        case D4LKeyTypes.APP_PUBLIC_KEY:
+            return importPublicKeyFromSPKI(key.pub);
+        case D4LKeyTypes.MAIN_KEY:
+            return importSymKeyFromHexadecimal(key.sym);
+        default:
+            throw new Error('invalid key type');
+    }
+};
+
+// GENERATE
+
+/**
+ * Creates a random symmetric key.
+ *
+ * @param {D4LKeyTypes} type of the key
+ * @param {AesKeyGenParams | HmacKeyGenParams | Pbkdf2Params} Cryptographic algorithm used.
+ * @returns {Promise<D4LKey>} Resolves to a symmetric key as a d4lKey object
+ */
+const generateSymKey = (type: D4LKeyTypes, algorithm: AesKeyGenParams | HmacKeyGenParams | Pbkdf2Params = AES_GCM): Promise<D4LKey> =>
+    // Parameters:
+    // 1. Symmetric Encryption algorithm name and its requirements
+    // 2. Boolean indicating extractable. which indicates whether or not the raw keying
+    //    material may be exported by the application
+    //    (http://www.w3.org/TR/WebCryptoAPI/#dfn-CryptoKey-slot-extractable)
+    // 3. Usage of the key. (http://www.w3.org/TR/WebCryptoAPI/#cryptokey-interface-types)
+    crypto.subtle.generateKey(
+        algorithm,
+        true,
+        [ENCRYPT, DECRYPT],
+    )
+        .then(key => exportKey(key, type));
+
+
+/**
+ * Creates key out of the given buffer (aka password)
+ *
+ * @param {Uint8Array} masterKey
+ * @param {Uint8Array} [salt] - The salt used for deriving, zeroed if not provided
+ * @param {AesDerivedKeyParams} algorithm - The algorithm used for deriving. AES GCM by default
+ * @returns {Promise<CryptoKey | D4LKey>} Resolves to a CryptoKey derived from the masterKey
+ */
+const deriveKey = (masterKey: Uint8Array, salt: Uint8Array = new Uint8Array(16), algorithm: AesDerivedKeyParams = AES_GCM): Promise<CryptoKey | D4LKey> => {
+    const ITERATIONS = 10000;
+
+    if (isEdge) {
+        return new Promise((resolve, reject) =>
+            // promisify
+            pbkdf2.pbkdf2(
+                convertArrayBufferViewToString(masterKey),
+                Buffer.from(salt),
+                ITERATIONS,
+                32,     // 32 bytes = 256 bits = key size of AES 256, the mode of operation is not relevant
+                'sha256',
+                (err, key) => {
+                    if (err) {
+                        reject(err);
+                        return;
+                    }
+
+                    resolve(key);
+                },
+            ),
+        )
+            .then(convertArrayBufferViewToBase64)
+            .then(importSymKeyFromBase64);
+    }
+
+    return crypto.subtle.importKey(
+        'raw',
+        masterKey,
+        { name: 'PBKDF2' },
+        false,
+        ['deriveKey', 'deriveBits'],
+    )
+        .then(key =>
+            crypto.subtle.deriveKey(
+                {
+                    name: 'PBKDF2',
+                    salt,
+                    iterations: ITERATIONS,
```

Comment:
Nit: Makes me wonder if it makes sense to have `ITERATIONS` is too generic as a name. Maybe defining a whole PBKDF2 object would be cool... or not? :smile:  Like in a JS file where you have a quick overview of all the ciphers we use... and maybe an indirection on top so we can easily... sorry... drifting into symbolic links :laughing:

Reply:
yeah, makes sense. Will reconsider the whole const suggestions altogether in a future PR.